code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.kt

```diff
@@ -0,0 +1,279 @@
+                        .build()
+                } else {
+                    if (schema != OptimizeConstants.JsonValues.SCHEMA_TARGET_DEFAULT) {
+                        Log.debug(
+                            OptimizeConstants.LOG_TAG,
+                            SELF_TAG,
+                            "Cannot create Offer object, provided data Map doesn't contain valid item data."
+                        )
+                        return null
+                    }
+                    Log.trace(
+                        OptimizeConstants.LOG_TAG,
+                        SELF_TAG,
+                        "Received default content proposition item, Offer content will be set to an empty string."
+                    )
+                    return Builder(id, OfferType.UNKNOWN, "").build()
+                }
+            } catch (e: ClassCastException) {
+                Log.warning(OptimizeConstants.LOG_TAG, SELF_TAG, "Cannot create Offer object, provided data contains invalid fields.")
+                null
+            } catch (e: DataReaderException) {
+                Log.warning(OptimizeConstants.LOG_TAG, SELF_TAG, "Cannot create Offer object, provided data contains invalid fields.")
+                null
+            }
+        }
+
+        /**
+         * Creates a Map<String, Any> using this Offer's attributes.
+         *
+         * @return Map<String, Any> containing Offer data.
+         */
+        @JvmStatic
+        fun Offer.toEventData(): Map<String, Any> {
+            val offerMap = mutableMapOf<String, Any>()
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ID] = id
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ETAG] = etag
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCORE] = score
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCHEMA] = schema
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_META] = meta
+
+            val data = mutableMapOf<String, Any>()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_ID] = id
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_TYPE] = type.toString()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT] = content
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_LANGUAGE] = language
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CHARACTERISTICS] = characteristics
+
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA] = data
+
+            return offerMap
+        }
+
+        private fun getContentFromOfferData(offerData: Map<String, Any?>): String {
+            val offerContent = offerData[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT]
+            return when (offerContent) {
+                is String -> offerContent
+                is Map<*, *> -> JSONObject(offerContent).toString()
+                else -> ""
+            }
+        }
+
+        fun List<Offer>?.displayed() {
```

Comment:
This extension function being places inside the offer class is very confusing. It don’t directly  act on a single Offer instance but rather a list of them. Could we please move it to a different place? Is there any reason to not having it in `Messaging.java` since its a public API?

Reply:
Same comment for `List<Offer>?.generateDisplayInteractionXdm()`

Reply:
Also could you please explain why you chose to make these extension functions? Wouldn't it be easier for consumers if it was an API which accepted a list of offers?